File: Additive_Bachelier_Option_Pricing_MATLAB/Bootstrap/bootstrap.m
function [disc, fwd_prices] = bootstrap(optionPrices, strikes)
    % This function calculates discount factors and forward prices from option price data
    %
    % Inputs:
    %   optionPrices: Matrix of option prices (rows = maturities, columns = strikes)
    %   strikes:      Vector of strike prices corresponding to the columns of optionPrices
    %
    % Outputs:
    %   disc:         Vector of discount factors for each maturity
    %   fwd_prices:   Vector of forward prices for each maturity
    
    % Get dimensions of the input matrix
    N = size(optionPrices);
    
    % Calculate number of discount factors (maturities)
    % Each maturity has 2 rows
    disc_num = N(1)/2;
    
    % Initialize output vectors
    disc = ones(disc_num, 1);        % Discount factors (first one is 1.0 at t=0)
    fwd_prices = zeros(disc_num-1, 1); % Forward prices
    
    % Loop through each maturity (starting from the second one)
    for i = 2:disc_num
        % Initialize empty matrix to store valid option price differences and strikes
        matrix = [];
        
        % Loop through each strike price
        for j = 1:N(2)
            % Check if both option prices (call and put) for this maturity and strike are valid
            if ~isnan(optionPrices(2*i-1, j)) && ~isnan(optionPrices(2*i, j))
                % Store the difference between the two option prices and the strike
                matrix = [matrix; optionPrices(2*i-1, j) - optionPrices(2*i, j), strikes(j)];
            end
        end
        
        % Calculate means for both the sinthetic forwards and strikes
        G_mean = mean(matrix(:, 1));
        K_mean = mean(matrix(:, 2));
        
        % Calculate the slope of the regression line using covariance/variance formula
        % This estimates the discount factor as the negative of the slope
        numerator = (matrix(:, 1) - G_mean)' * (matrix(:, 2) - K_mean);
        denominator = sum((matrix(:, 2) - K_mean).^2);
        disc(i) = -numerator / denominator;
       
        par = [matrix(:, 2) ones(size(matrix(:, 2)))] \ matrix(:, 1);
        
        % Calculate forward price from regression intercept and discount factor
        fwd_prices(i-1) = par(2) / disc(i);
    end
end

File: Additive_Bachelier_Option_Pricing_Python/Bootstrap.py
import pandas as pd
import numpy as np
from pathlib import Path
from datetime import datetime
import matplotlib.pyplot as plt


def get_dates():
    """Returns futures expiration dates as datetime objects."""
    return pd.to_datetime([
        '2020-06-02', '2020-08-17', '2020-11-17', '2021-02-17',
        '2021-05-17', '2021-08-17', '2021-11-16', '2022-05-17',
        '2022-11-16'
    ])

def read_csv_data(file1, file2, target):
    """Reads and filters data for a given target date, returns prices as numpy array."""
    df1 = pd.read_csv(file1)
    df2 = pd.read_csv(file2)

    filtered1 = df1[df1.iloc[:, 0] == target].iloc[:, 1:]
    filtered2 = df2[df2.iloc[:, 0] == target].iloc[:, 1:]

    return np.vstack([filtered1.values, filtered2.values])

def build_option_prices(call_folder, put_folder, target):
    """Builds option price matrix and strike vector from calls and puts directories."""
    call_folder = Path(call_folder)
    put_folder = Path(put_folder)

    call_files = sorted(call_folder.glob('*call.csv'))
    if not call_files:
        raise FileNotFoundError(f'No "*call.csv" files found in {call_folder}')

    # Use first file to get strikes
    first_df = pd.read_csv(call_files[0])
    strikes = first_df.columns[1:].astype(float).values
    N = len(strikes)
    M = len(call_files)

    option_prices = np.zeros((2 * M, N))

    for k, call_file in enumerate(call_files):
        put_file = put_folder / call_file.name.replace('call', 'put')
        if not put_file.exists():
            raise FileNotFoundError(f'Missing put file: {put_file}')

        block = read_csv_data(call_file, put_file, target)
        option_prices[2 * k: 2 * k + 2] = block

    return option_prices, strikes

def bootstrap(option_prices, strikes):
    """
    Bootstrap discount factors and forward prices from option prices.

    Parameters:
        option_prices (ndarray): Matrix of option prices (2M x N)
        strikes (ndarray): Vector of strike prices (length N)

    Returns:
        disc (ndarray): Vector of discount factors (length M)
        fwd_prices (ndarray): Vector of forward prices (length M-1)
    """
    M = option_prices.shape[0] // 2
    N = option_prices.shape[1]

    disc = np.ones(M)
    fwd_prices = np.zeros(M - 1)

    for i in range(1, M):  # Start from the second maturity
        call_row = option_prices[2 * i]
        put_row = option_prices[2 * i + 1]

        call_put_diff, K = zip(*[
        (c - p, k) for c, p, k in zip(call_row, put_row, strikes)
        if np.isfinite(c) and np.isfinite(p)
        ])
        call_put_diff = np.array(call_put_diff)
        K = np.array(K)

        G_mean = np.mean(call_put_diff)
        K_mean = np.mean(K)

        # Calculate discount factor as negative of regression slope
        numerator = np.dot(call_put_diff - G_mean, K - K_mean)
        denominator = np.dot(K - K_mean, K - K_mean)
        disc[i] = - numerator / denominator

        # Linear regression: call_put_diff = a*K + b
        A = np.vstack([K, np.ones_like(K)]).T
        a, b = np.linalg.lstsq(A, call_put_diff, rcond=None)[0]

        # Forward price is intercept / discount
        fwd_prices[i - 1] = b / disc[i]

    return disc, fwd_prices

def yearfrac(t0, dates, basis=3):
    """
    Compute year fraction between t0 and each date in dates.

    Parameters:
        t0 (datetime): Starting date
        dates (Series or array): Target dates
        basis (int): Day count convention. 3 = ACT/365

    Returns:
        Array of year fractions
    """
    if basis == 3:
        return (dates - t0).days / 365.0
    else:
        raise NotImplementedError("Only ACT/365 (basis=3) is implemented")

def zero_rates(dates, discounts, t0):
    """
    Compute zero rates from discount factors and dates.

    Parameters:
        dates (Series): Dates corresponding to discount factors
        discounts (ndarray): Discount factors
        t0 (datetime): Reference (value) date

    Returns:
        Array of zero rates
    """
    deltas = yearfrac(t0, dates)
    return -np.log(discounts) / deltas

from datetime import datetime

def getDatesNew(t0):
    """
    Parameters:
    - t0: datetime.datetime, value date

    Returns:
    - list of datetime.datetime: vector of futures expiry dates
    """
    return [
        t0,
        datetime(2020, 8, 17),
        datetime(2020, 11, 17),
        datetime(2021, 2, 17),
        datetime(2021, 5, 17),
        datetime(2021, 8, 17),
        datetime(2021, 11, 16),
        datetime(2022, 5, 17),
        datetime(2022, 11, 16)
    ]

